src/Type/Set/VariantF.hs

```diff
@@ -0,0 +1,194 @@
+  forMember :: (Follow ss bst ~ f) =>
+    SSide ss -> Dict (c (Follow ss bst))
+
+instance (c a, ForAllIn c l, ForAllIn c r
+         ) => ForAllIn (c :: k -> Constraint) ('Branch (a :: k) (l :: TypeSet k) (r :: TypeSet k)) where
+
+  -- forMember :: forall (ss :: [Side]). SSide ss -> Dict (c (Follow ss ('Branch a l r)))
+  forMember SNil = (Dict :: Dict (c a))
+  forMember (SL ss)
+    = case proveFollowLeft of
+        HRefl -> forMember @_ @c @l ss-- :: Dict (c (Follow ss l)))
+  forMember (SR ss)
+    = case proveFollowRight of
+        HRefl -> forMember @_ @c @r ss-- :: Dict (c (Follow ss r)))
+
+instance () => ForAllIn (c :: k -> Constraint) 'Empty where
+  forMember = error "Somehow got invalid path into TypeSet"
+
+
+type family ForAll (c :: k -> Constraint) (bst :: TypeSet k) :: Constraint
+
+------------------------------------------------------------------------------
+-- | A 'VariantF' is a higher-order version of 'Variant' which can contain
+--   any of the 'Functor's within its 'TypeSet'. You can use 'toVariantF' to
+--   construct one, and 'fromVariantF' to pattern match it out.
+data VariantF (bst :: TypeSet (* -> *)) (a :: *) where
+  VariantF :: ()
+    => SSide ss -> Follow ss bst a -> VariantF bst a
+type role VariantF nominal nominal
+
+
+------------------------------------------------------------------------------
+-- | A proof that the set @bst@ contains functor @f@.
+class HasF f bst where
+
+  -- | Inject a @t@ into a 'VariantF'.
+  toVariantF :: f a -> VariantF bst a
+
+  -- | Attempt to project a 'VariantF' into @f a@. This might fail, because there
+  -- is no guarantee that the 'VariantF' /actually contains/ @f a@.
+  fromVariantF :: VariantF bst a -> Maybe (f a)
+
+instance ( Follow (Locate f bst) bst ~ f
+         , FromSides (Locate f bst)
+         ) => HasF f bst where
+  toVariantF = VariantF (fromSides @(Locate f bst))
+  fromVariantF (VariantF tag res) =
+    case testEquality tag (fromSides @(Locate f bst)) of
+      Just Refl -> Just res
+      Nothing -> Nothing
+
+-- forall c f bst. (ForAll c bst, HasF f bst) => c f
+
+------------------------------------------------------------------------------
+-- | TODO
+data SplitF f lbst rbst a
+  = RootF (f a)
+  | LSplitF (VariantF lbst a)
+  | RSplitF (VariantF rbst a)
+
+------------------------------------------------------------------------------
+-- | TODO
+decompRootF :: VariantF ('Branch f lbst rbst) a -> SplitF f lbst rbst a
+decompRootF (VariantF SNil   t) = RootF t
+decompRootF (VariantF (SL s) t) = LSplitF (VariantF s t)
+decompRootF (VariantF (SR s) t) = RSplitF (VariantF s t)
+
+------------------------------------------------------------------------------
+-- | A proof that inserting into a @bst@ doesn't affect the position of
+-- anything already in the tree.
+proveFollowInsertF :: forall (ss :: [Side]) (f :: * -> *) (bst :: TypeSet (* -> *)).
+  Follow ss (Insert f bst) :~~: Follow ss bst
+proveFollowInsertF = unsafeCoerce HRefl
```

Comment:
Is this different from the one in `Type.Set.Variant`? Can we reuse the same one?

Reply:
Technically yes it's using `:~~:` which is an equality proof for arbitrary kinds instead of `:~:` which seems to only work on the types with kind `*`. 

I'm pretty confident it can be replaced, but I was trying to keep the changes localized so it would be easier for you to review and because I'm not sure how you'd want it to be organized.